torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1210,6 +1233,14 @@ def __init__(
         # range may contain ints which may not actually appear in
         # practice
         self.var_to_range: Dict["sympy.Symbol", ValueRanges] = {}
+        # var_to_range entries which were modified by manual user directive
+        # [NOTE - on user_constrained]
+        # User constrained symbols have an entry in var_to_range, but some of their range
+        # comes from manual user directives like dynamo's constrain api. This distinction is maintained
+        # for the purposes of protecting and enforcing user directives. In strict_mark_dyn mode, we do not
+        # allow ANY constraining of values for a given symbol. However, if this symbol is user constrained, we
+        # relax this requirement in favor of honoring the user directive and allowing the user specified range.
+        self.user_constrained: Set["sympy.Symbol"] = set()
```

Comment:
I don't understand. All uses of `constrain_range` can only occur from "users". We do call constrain_range in nnz but this is to setup a "common sense" constrain range on unbacked symint. There are no uses of `constrain_range` otherwise.

Reply:
We still need a structure to check, because every single symbol gets the "common sense" constraint. The presence of an external user invoked constrain changes the semantics of export + mark_dynamic.

Export + mark_dynamic --> no constraining allowed
Export + mark_constrain_dynamic --> constraining allowed within a range

Open to other ideas for doing this, but this was the simplest way WITHOUT checking if the range == "common sense" range. I figured just checking the "common sense" range was kinda grosser than this.